source/core/Events.cpp

```diff
@@ -0,0 +1,776 @@
+
+        UInt32 msTimeout = timeOutPtr ? *timeOutPtr : -1;
+        VIClump* clump = THREAD_CLUMP();
+        if (!occ.HasOccurred(staticCount, false)) {
+            Observer* pObserver = clump->GetObservationStates(2);
+            if (!pObserver) {
+                PlatformTickType future = msTimeout > 0 ? gPlatform.Timer.MillisecondsFromNowToTickCount(msTimeout) : 0;
+                pObserver = clump->ReserveObservationStatesWithTimeout(2, future);
+                occ.InsertObserver(pObserver+1, occ.Count()+1);
+                return clump->WaitOnObservableObject(_this);
+            }
+        } else {
+            // gPlatform.IO.Printf("Event Has Occurred %d %d\n", staticCount, occ.Count());
+        }
+        clump->ClearObservationStates();
+        ++staticCount;
+    }
+
+    // TODO(spathiwa) -- watch static and dynamic queue, take earliest event
+    // TODO(spathiwa) -- handle/test for spurious timeout (real event and timeout happen at same time,
+    //                   should go back to sleep w/o running timeout case)
+    const EventData &eventData = EventOracle::TheEventOracle().GetEventData(eventQID);
+    Int32 dynIndex = regInfo ? regInfo->DynamicEventMatch(eventData.common) : 0;
+
+    for (Int32 inputTuple = 0; inputTuple < numTuples; ++inputTuple) {
+        UInt32 eventSpecIndex = *arguments[inputTuple].eventSpecIndex;
+        if (eventSpecIndex >= esCount) {
+            THREAD_EXEC()->LogEvent(EventLog::kHardDataError, "Invalid Event Spec Index");
+            return THREAD_EXEC()->Stop();
+        }
+        TypeRef esEventDataNodeType = arguments[inputTuple].data._paramType;
+        void *esEventDataNode = arguments[inputTuple].data._pData;
+        InstructionCore *esBranchTarget = arguments[inputTuple].branchTarget;
+        if (EventMatch(eventData, eventSpecRef[eventSpecIndex], dynIndex)) {
+            // gPlatform.IO.Printf("Event match %d -> %x\n", eventSpecRef[eventSpecIndex].eventType, esBranchTarget);
+            const Int32 commonDataSize = sizeof(EventCommonData);
+            Int32 dataNodeSize = esEventDataNodeType->TopAQSize();
+            if (dataNodeSize >= commonDataSize) {
+                // TODO(spathiwa) Compute eventIndex when we support multiple events sharing event cases
+                UInt32 eventIndex = 0;
+                memcpy(esEventDataNode, &eventData, commonDataSize);
+                *EventIndexFieldPtr(esEventDataNode) = eventIndex;
+
+                if (eventData.eventDataType->TopAQSize() == dataNodeSize - commonDataSize) {
+                    // TODO(spathiwa) Should we also verify type of event data node matches event at run-time?
+                    // When we support multiple events with a shared event case, this needs to be smarter
+                    // and copy only common data.
+                    eventData.eventDataType->CopyData(eventData.pEventData, (AQBlock1*)esEventDataNode + commonDataSize);
+                }
+            } else {
```

Comment:
I would add a comment that this is for the timeout case.

Reply:
Done.